### PR TITLE
Add key extension logic

### PR DESCRIPTION
**Pull request by rowannekabalan**
## What does this change?
This moves the developer key extension/deletion logic from the User level to the Key level. This was done to avoid deleting a user and all their keys. Once a user doesn't have any keys left they will be deleted automatically.

## How to test
To be end-to-end tested on CODE

## How can we measure success?
Users should be able to extend/delete their developer keys.

## Have we considered potential risks?
This is a major change to the way we handle key cleanup logic, so we need to make sure we didn't break anything.

### CODE REVIEW

File: app/store/Dynamo.scala [hunk 1]
```diff
@@ -179,6 +181,14 @@ class Dynamo(db: DynamoDB, usersTable: String, keysTable: String, labelTable: St
       new AttributeUpdate("requests_per_minute").put(kongKey.requestsPerMinute),
       new AttributeUpdate("status").put(kongKey.status),
       new AttributeUpdate("tier").put(kongKey.tier.toString),
+      kongKey.remindedAt match {
+        case Some(date) => new AttributeUpdate("extendedAt").put(date)
```
Comment (alinaboghiu) [inline app/store/Dynamo.scala:185]:
```suggestion
        case Some(date) => new AttributeUpdate("remindedAt").put(date)
```

File: app/store/Dynamo.scala [hunk 2]
```diff
@@ -179,6 +181,14 @@ class Dynamo(db: DynamoDB, usersTable: String, keysTable: String, labelTable: St
...
+        case None => new AttributeUpdate("extendedAt").delete()
```
Comment (alinaboghiu) [inline app/store/Dynamo.scala:186]:
```suggestion
        case None => new AttributeUpdate("remindedAt").delete()
```

File: app/store/Dynamo.scala [hunk 3]
```diff
@@ -416,6 +429,9 @@ object Dynamo {
       .withList("labelIds", labelIds.asJava)
 
     kongKey.productUrl.foreach(productUrl => item.withString("productUrl", productUrl))
+    kongKey.remindedAt.foreach(remindedAt => item.withLong("remindedAt", remindedAt))
+    kongKey.extendedAt.foreach(remindedAt => item.withLong("remindedAt", remindedAt))
```
Comment (alinaboghiu) [inline app/store/Dynamo.scala:433]:
```suggestion
    kongKey.extendedAt.foreach(extendedAt => item.withLong("extendedAt", remindedAt))
```

File: app/logic/DeveloperFormLogic.scala [hunk 4]
```diff
@@ -43,26 +43,30 @@ class DeveloperFormLogic(dynamo: DB, kong: Kong) {
...
+
+  def deleteKeyByUser(key: KongKey): Future[_] = {
+    Future {
+      for {
+        _ <- kong.deleteKey(key.kongConsumerId).recover {
+          case Kong.GenericFailure(txt) if txt.contains("404") => ()
+        }
+      } yield {
+        dynamo.deleteKey(key)
+        if (dynamo.getKeysWithUserId(key.bonoboId) == 0) {
```
[the remark this answers is not in the document]
Comment (rowannekabalan) [reply app/logic/DeveloperFormLogic.scala:63]: Ah good catch thank you!

**Review by alinaboghiu — APPROVED** [review]

**Pull request merged**